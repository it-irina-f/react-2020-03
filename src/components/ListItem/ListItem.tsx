import React from "react";
import { IconEdit, IconTrash2 } from "sancho";
import styled from "@emotion/styled";
import { ManageButton } from "@/components/Buttons";

interface ListItemProps {
  id: number;
  text: string;
  isComplete: boolean;
}

type TypeIdNumber = (id: number) => void;

interface Props {
  listItem: ListItemProps;
  toggleComplete: TypeIdNumber;
  deleteListItem: TypeIdNumber;
  handleEdit: TypeIdNumber;
}

const ListItemWrapper = styled.div`
  display: flex;
  margin: 10px auto;
  border-bottom: 1px solid #ddd;
  padding: 0 0 10px;
`;

const LabelWrapper = styled.label`
  flex-grow: 1;
`;

export const ListItem: React.FC<Props> = ({
  listItem,
  toggleComplete,
  deleteListItem,
  handleEdit,
}) => {
  return (
    <ListItemWrapper>
      <LabelWrapper>
        <input
          type="checkbox"
          name={"checkbox_" + listItem.id}
          checked={listItem.isComplete}
          onChange={() => toggleComplete(listItem.id)}
        />
        {listItem.text}
      </LabelWrapper>
      <ManageButton
        icon={<IconEdit />}
        onClick={() => handleEdit(listItem.id)}
        label="editListItem"
        id={"editListItem_" + listItem.id}
      />
      <ManageButton
        icon={<IconTrash2 />}
        onClick={() => deleteListItem(listItem.id)}
        label="deleteListItem"
        id={"deleteListItem_" + listItem.id}
      />
    </ListItemWrapper>
  );
};
