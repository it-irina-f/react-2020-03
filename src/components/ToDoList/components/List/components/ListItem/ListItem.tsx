import React from "react";

import type { ListItemProps, ToggleCompleteProps } from "types/todo";

interface Props {
  listItem: ListItemProps;
  toggleComplete: ToggleCompleteProps;
}

export const ListItem: React.FC<Props> = ({ listItem, toggleComplete }) => {
  return (
    <div>
      <label>
        <input
          type="checkbox"
          name={"checkbox_" + listItem.id}
          checked={listItem.isComplete}
          onChange={() => toggleComplete(listItem.id)}
        />
        {listItem.text}
      </label>
    </div>
  );
};
