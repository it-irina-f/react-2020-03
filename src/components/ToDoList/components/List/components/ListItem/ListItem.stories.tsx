import React from "react";
import { action } from "@storybook/addon-actions";
import { withKnobs, object } from "@storybook/addon-knobs";
import { ListItem } from "./ListItem";

export default {
  title: "ListItem Component",
  decorators: [withKnobs],
};

const listItem1 = { id: 0, text: "Полить цветы", isComplete: true };
const listItem2 = { id: 1, text: "Сделать ДЗ", isComplete: false };

export const noComplete = () => [
  <ListItem
    listItem={object("listItem", { ...listItem1 })}
    toggleComplete={action("Item clicked")}
    deleteListItem={action("Click for delete item")}
    editListItem={action("Click for edit item")}
    key="jsx"
  ></ListItem>,
];

export const complete = () => [
  <ListItem
    listItem={object("listItem", { ...listItem2 })}
    toggleComplete={action("Item clicked")}
    deleteListItem={action("Click for delete item")}
    editListItem={action("Click for edit item")}
    key="jsx"
  ></ListItem>,
];
