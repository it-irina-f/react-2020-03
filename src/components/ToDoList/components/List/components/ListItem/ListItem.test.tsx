import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";

import { ListItem } from "./ListItem";

const listItem1 = { id: 0, text: "Полить цветы", isComplete: true };
const listItem2 = { id: 1, text: "Сделать ДЗ", isComplete: false };

const click = jest.fn();

describe("Component ListItem", () => {
  it("renders list item where isComplete=true", () => {
    expect(
      renderer
        .create(
          <ListItem
            listItem={listItem1}
            toggleComplete={click}
            deleteListItem={click}
            handleEdit={click}
          />
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  it("renders list item where isComplete=false", () => {
    expect(
      renderer
        .create(
          <ListItem
            listItem={listItem2}
            toggleComplete={click}
            deleteListItem={click}
            handleEdit={click}
          />
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  it("simulate onChange checkbox by list item where isComplete=true", () => {
    const wrapper = mount(
      <ListItem
        listItem={listItem1}
        toggleComplete={click}
        deleteListItem={click}
        handleEdit={click}
      />
    );
    wrapper.find("input").simulate("change");
    expect(click).toHaveBeenCalled();
  });
  it("simulate onChange checkbox by list item where isComplete=false", () => {
    const wrapper = mount(
      <ListItem
        listItem={listItem2}
        toggleComplete={click}
        deleteListItem={click}
        handleEdit={click}
      />
    );
    wrapper.find("input").simulate("change");
    expect(click).toHaveBeenCalled();
  });

  it("should be able to click the button Delete Item", () => {
    const wrapper = mount(
      <ListItem
        listItem={listItem2}
        toggleComplete={click}
        deleteListItem={click}
        handleEdit={click}
      />
    );
    wrapper.find("button[name='deleteListItem']").simulate("click");
    expect(click).toHaveBeenCalled();
  });

  it("should be able to click the button Edit Item", () => {
    const wrapper = mount(
      <ListItem
        listItem={listItem2}
        toggleComplete={click}
        deleteListItem={click}
        handleEdit={click}
      />
    );
    wrapper.find("button[name='editListItem']").simulate("click");
    expect(click).toHaveBeenCalled();
  });
});
