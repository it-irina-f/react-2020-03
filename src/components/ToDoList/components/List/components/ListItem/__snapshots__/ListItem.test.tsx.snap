// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component ListItem renders list item where isComplete=false 1`] = `
<div>
  <label>
    <input
      checked={false}
      name="checkbox_1"
      onChange={[Function]}
      type="checkbox"
    />
    Сделать ДЗ
  </label>
  <button
    name="editListItem"
    onClick={[Function]}
    type="button"
  >
    Редактировать
  </button>
  <button
    name="deleteListItem"
    onClick={[Function]}
    type="button"
  >
    Удалить
  </button>
</div>
`;

exports[`Component ListItem renders list item where isComplete=true 1`] = `
<div>
  <label>
    <input
      checked={true}
      name="checkbox_0"
      onChange={[Function]}
      type="checkbox"
    />
    Полить цветы
  </label>
  <button
    name="editListItem"
    onClick={[Function]}
    type="button"
  >
    Редактировать
  </button>
  <button
    name="deleteListItem"
    onClick={[Function]}
    type="button"
  >
    Удалить
  </button>
</div>
`;
